RetroAchievements/RAWeb
# feat(game-list): redesign the ui for mobile devices

This PR redesigns the React game list datatable UI for mobile devices. The ultimate goals are to:

* Eliminate all horizontal scrolling.
* Simplify searching, filtering, sorting, and pagination.
* Not impact the desktop UX in any way.

## How to test this PR
```bash
sail npm install
sail npm run build
sail artisan inertia:start-ssr
```
Navigate to `/games` and/or `/game-list/play`.

Shrinking your browser window to XS won't be sufficient. The page is explicitly checking for a mobile user agent to cut down on the JS sent to the browser and improve performance.

In your browser devtools, you'll need to explicitly simulate a mobile device. Click the highlighted button, and then ensure you're set to iPhone SE:

![Screenshot 2024-10-14 at 3 40 00 PM](https://github.com/user-attachments/assets/21adfc12-1859-406e-a62f-690be6229424)

![Screenshot 2024-10-14 at 3 41 18 PM](https://github.com/user-attachments/assets/c7509cf7-b9c7-45e0-8b8a-c8b7ef1f95b9)

Be sure to refresh the page after you're set to iPhone SE.

## UX Walkthrough

### Before and After

![Screenshot 2024-10-14 at 3 50 21 PM](https://github.com/user-attachments/assets/5ba097e3-9c18-4a21-9810-d768e92babe4)

* There is no table in the DOM. Things now render in a touch-oriented list.
* The mobile UX is modeled closely after music apps such as Apple Music and Spotify. Above the list, we have a count of games, a "super filter button", and a searchbar.
* Game images are bigger, and titles are truncated to a single line of text. The system is always shown under the title. Other information can appear next to the system chip based on the user's current context.
* Each row has two actions: backlog toggle and open more details.

### The Super Filter Button

All filters and sort orders are wrapped into a single button which pops a drawer onto the screen:

https://github.com/user-attachments/assets/459354db-7636-46ec-8eda-62c4742c4e40

Tracking has been added to sort orders on both desktop and mobile. I highly suspect there are a few of these sort orders which never actually get used. After we have some data, we can organize these a little more intelligently in the list and possibly prune sort orders that are bloating the UX.

When the user changes their sort order, their current sort context always appears on the list items:

https://github.com/user-attachments/assets/cf26de17-c83f-49d8-8e4c-0b2ca3439494

### Backlog Toggle

The backlog toggle control works just like it does on desktop. It's worth noting that the word "playlist" is used instead of "Want to Play Games List", due to space constraints. "Playlist" and "devlist" are terms that may be worth considering rebranding both of those features to in the future.

https://github.com/user-attachments/assets/4732ce76-b7b3-4a8b-96ff-e7d24cb4f847

### Game Details Button

Tapping the three vertical dots button on each element opens a drawer from the bottom of the page. This drawer is something of a mobile/touch alternative to the game tooltip cards that users with a mouse currently enjoy:

https://github.com/user-attachments/assets/8d72be97-5b8e-4b23-adc0-1a8cdc3cd5ef

![Screenshot 2024-10-14 at 4 24 53 PM](https://github.com/user-attachments/assets/3e98d15e-a597-456d-9a3d-3159acd8dc81)

There are a few nuances here that were worth pointing out:
* The progress list item is not displayed if the user is not authenticated.
* While the drawer is open, toggling the item on the backlog will _not_ pop a toast. Popping a toast while a drawer is open causes some bizarre UX consequences. Visual feedback is still provided: the + icon rotates like usual, and the button changes variant.
* There are three links to the game page in the drawer. The "Open Game" button is the obvious one. The other two are subtle: the user can tap the game avatar and the game title. These two subtle links don't really announce themselves with `text-link` and are just there to capture panic taps.
* If the user taps the backlog toggle while not authenticated, they're redirected to login.

### Simplified Pagination

I wanted to eliminate the mobile user from having to think about pagination as much as possible. I tried several different approaches, including infinite scroll, but I scrapped this after I wasn't too happy with how it tested. I landed somewhere in the middle between infinite scrolling and manual pagination:

https://github.com/user-attachments/assets/25911ee2-670e-4309-82da-aa937c1e5efa

There's some trickery happening here. When the user scrolls to within 300px of the "Load more" button, we kick off the fetch for the next page. When/if they tap the button, assuming they have a stable network connection, the results instantly display. This creates an illusion that the back-end is responding instantaneously, even though it most assuredly is not.

If the user has a slow network connection, the UI is smart enough to detect that the fetch is still in flight and shows a temporary loading state until the fetch has completed.

## Other things worth noting
* Mobile's default page size is 100, not 25.
* Landscape (widescreen / side-to-side orientation) mode is basically ignored just due to how little use it gets.
* A lot of test coverage has been added for all this functionality.
